gaeusersaccountmanager.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+)
+
+const emailHeaderKey = "X-Appengine-User-Email"
+const usernameHeaderKey = "X-Appengine-User-Nickname"
+
+type GAEUsersAccountManager struct{}
+
+func NewGAEUsersAccountManager() *GAEUsersAccountManager {
+	return &GAEUsersAccountManager{}
+}
+
+func (g *GAEUsersAccountManager) Authorize(fn AuthHTTPHandler) HTTPHandler {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		// These headers are guaranteed to be present and come from AppEngine.
```

Comment:
Would it be possible to use the official [appengine package](https://pkg.go.dev/google.golang.org/appengine) here? It includes a support library for [user information](https://pkg.go.dev/google.golang.org/appengine@v1.6.7/user).

Reply:
Yes, but I chose not to because of the following in no particular order:
- I don't want to make it too easy to add even more dependencies on app engine
- it's easier (in my opinion) to get the information from the headers than to create a dependency with a library and then use it
- we're not exposed to bugs or security issues from the library (unlikely to affect us for such a simple use, but not impossible)
- if app engine were to do a breaking change (change the header or stop using it altogether) we'd still be affected by it because we'd need to update the version of the library

Reply:
What I'm afraid of here is rewriting the app engine package through a series of "one more thing" changes. If the issue is that you want to keep the option open for non-appengine implementations of AccountManager, would it be too much additional work to figure out what the equivalent would be for e.g. AWS to validate that this is extensible?

Reply:
This code is not trying to be extensible, it's doing a very app engine specific thing.

"rewriting the app engine package through a series of "one more thing" changes"
That sounds bad, but is it really?
- The header approach is the only option for some runtimes (example nodejs), so it's unlikely to change.
- The code for getting the username from the app is not going to be simpler than this (call a function versus get a value from a dictionary) and will require additional configuration and dependency handling.

If another dependency on app engine needs to be added later we can decide whether it makes sense to add the library then or go this route again.